### ISSUE TITLE
debug-adapter: no way to change the PC

### ISSUE THREAD

**Issue by maziac** [Microsoft/vscode#51716]
Hi,
with my debug adapter I have the next problem.
I would like to change the PC (Program Counter) in the debugger from the console (or from a menu).
How do I get vscode updated to reflect the new PC.
I.e. is there any event I can send to vscode so that vscode changes the PC-cursor to the line in the source file.

The best I could find so far is to send a  StoppedEvent, e.g.
~~~
this.sendEvent(new StoppedEvent('step', EmulDebugAdapter.THREAD_ID));
~~~

This result in vscode requesting the callstack once again and I can at least reflect the new state in the callstack.
However the visual cursor still points to the old source line.

**Comment by yume-chan** [comment]
Like a custom `execute to here` command?

**Comment by maziac** [comment]
No, it's not about 'next'.

In Visual Studio (not vscode) for example it is possible to drag the PC to another line in the source code. Afterwards you can continue from that source code line.

I'm trying to mimic something similar here and added a menu like this:

![image](https://user-images.githubusercontent.com/3081973/41362290-465479a2-6f31-11e8-9e38-328555488355.jpeg)

When the user selects this the PC will be changed in this example to line 3570 which is address 0x0bce.
But the yellow cursor does not move.

**Comment by weinand** [comment]
Using the stopped event is the correct approach.
With this VS Code will ask for the stack and the top stack frame's "line" location is used to adjust the yellow arrow.

BTW, the Debug Adapter Protocol has support for a "Goto" request:
https://github.com/Microsoft/vscode-debugadapter-node/blob/f38fbd3d172288722ef1143ec5f43bfbbe01283d/protocol/src/debugProtocol.ts#L650-L671
and 
https://github.com/Microsoft/vscode-debugadapter-node/blob/f38fbd3d172288722ef1143ec5f43bfbbe01283d/protocol/src/debugProtocol.ts#L650-L671

but VS Code does not (yet) provide a UI for it. But it is on our 6 Month road map.
It would be great if your debug extension could implement the two Goto requests (even if you have to implement the UI for it yourself).

**Comment by maziac** [comment]
That is great news. I'm happy to implement it when it's ready.

If the StoppedEvent is the right thing to use what should I use as reason. 'step' doesn't seem to be right. BTW what is the reason used for ?

And another problem: This is exactly as I use it right now. Unfortunately the PC cursor is not moved until I click on the topmost call stack item.

**Comment by weinand** [comment]
The StoppedEvent's reason should by "stop" (or whatever should be shown in the UI).
See https://github.com/Microsoft/vscode-debugadapter-node/blob/f38fbd3d172288722ef1143ec5f43bfbbe01283d/protocol/src/debugProtocol.ts#L78-L84

VS Code does not interpret the reason in any way.

> And another problem: This is exactly as I use it right now. Unfortunately the PC cursor is not moved until I click on the topmost call stack item.

Most likely you need to send a Continue event so that VS Code understands that the program has resumed (even if this is not completely true in the Goto case).
See https://github.com/Microsoft/vscode-debugadapter-node/blob/f38fbd3d172288722ef1143ec5f43bfbbe01283d/protocol/src/debugProtocol.ts#L99-L112

**Comment by maziac** [comment]
StoppedEvent-reason: Now I got it. I was always wondering where this reason goes to. It took me a while but after heavy searching I found the UI element in the CALL STACK. OK, so it's just displayed.

Continued-Event: I tried it and it works fine. The PC cursor is moved. Thanks again.
I have seen this initiates a scopes etc. request. Is this maybe also the answer to my other question:
https://github.com/Microsoft/vscode/issues/49008
